test-validator/src/lib.rs

```diff
@@ -312,6 +313,38 @@ impl TestValidatorGenesis {
         Ok(self)
     }
 
+    pub fn clone_upgradeable_programs<T>(
+        &mut self,
+        addresses: T,
+        rpc_client: &RpcClient,
+    ) -> Result<&mut Self, String>
+    where
+        T: IntoIterator<Item = Pubkey>,
+    {
+        let addresses: Vec<Pubkey> = addresses.into_iter().collect();
+        self.clone_accounts(addresses.clone(), rpc_client, false)?;
+
+        let mut programdata_addresses: HashSet<Pubkey> = HashSet::new();
+        for address in addresses {
+            let account = self.accounts.get(&address).unwrap();
+
+            if let Ok(UpgradeableLoaderState::Program {
+                programdata_address,
+            }) = account.deserialize_data()
```

Comment:
Since the programdata address is deterministic from `address` and the loader, I don't think we need to deserialize `account` at all, right?
I think this could be:
```
let (programdata_address, _) = Pubkey::find_program_address(&[address.as_ref()], &account.owner);
```

Reply:
You're probably right, but we can be 100% sure that this implementation is always correct, *just in case* the program data account is ever on a different address.

For example, if we decide to make the token program upgradeable, and go through our normal upgrade process of replacing the account, we could break that determinism. Just for that eventuality, I would prefer to keep this implementation. It's a bit of a stretch, since we would probably do that upgrade more carefully, so I could be convinced to change.

It does beg the question though: why does the program account store this data at all?

Reply:
> It does beg the question though: why does the program account store this data at all?

Yeah, good question! We might be able to save 32 bytes per program account.